### PR TITLE
Playwright_PEPPER-1299_Verify OS2 realm webelements in DSM

### PR DESCRIPTION
**Pull request by Rinchan20**
PR is meant to verify as much as possible related to webelements in OS PE-CGS/OS2 realm in DSM (ticket: https://broadworkbench.atlassian.net/browse/PEPPER-1299)

likely to be combined with the ticket for checking the display of info for an OS1 -> OS2 participant: https://broadworkbench.atlassian.net/browse/PEPPER-1300

### CODE REVIEW

File: playwright-e2e/dsm/enums.ts [hunk 1]
```diff
@@ -26,22 +26,47 @@ export enum DataFilter {
 // Table column headers
 
 export enum CustomizeView {
-  CONTACT_INFORMATION = 'Contact Information Columns',
```
Comment (Rinchan20) [inline playwright-e2e/dsm/enums.ts:29]: this is just changing things around in this enum to make it alphabetical - to make it easier to spot a missing CustomizeView column group

File: playwright-e2e/tests/dsm/osteo/os2-display-verification.spec.ts [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+    })
+
+    await test.step(`Verify that the Download List and Bulk Cohort Tagis displayed`, async () => {
+      await participantListPage.assertDownloadListDisplayed();
+      await participantListPage.assertBulkCohortTagDisplayed();
+    })
+
+    const customizeViewPanel = participantListPage.filters.customizeViewPanel;
+    await customizeViewPanel.open();
+
```
Comment (Rinchan20) [inline playwright-e2e/tests/dsm/osteo/os2-display-verification.spec.ts:140]: Checking all customize view options looks long - but only seems to take the test ~ 1 min, 50 seconds to check

File: playwright-e2e/dsm/component/tables/participant-list-table.ts [hunk 3]
```diff
@@ -75,6 +76,37 @@ export class ParticipantListTable extends Table {
     return cell.innerText();
   }
 
+  /**
+   * Use to verify that certain headers are currently displayed in the participant list
+   * @param checkDefaultFilterOfStudy choose whether to check for the default filter of a given study - use in combination with studyName param
+   * @param studyName the name of the study whose default filter should be checked
+   * @param customFilter for use to check display of custom filters - this is a given list of column headers that make up the given filter
+   */
+  public async assertDisplayedHeaders(opts: { checkDefaultFilterOfStudy?: boolean, studyName?: StudyName, customFilter?: Label[] }): Promise<void> {
```
Comment (aweng98) [inline playwright-e2e/dsm/component/tables/participant-list-table.ts:86]: nit: We ought to start writing code for business logics inside the test script. Writing business logic functions inside a page object file is bad practice.

File: playwright-e2e/dsm/pages/participant-list-page.ts [hunk 4]
```diff
@@ -54,10 +54,20 @@ export default class ParticipantListPage extends DsmPageBase {
     await checkbox.check({ timeout: 30 * 1000 }); // long delay
   }
 
+  public async assertSelectAllDisplayed(): Promise<void> {
+    const checkbox = this.page.getByRole('checkbox', { name: 'Select all' });
+    await expect(checkbox).toBeVisible();
+  }
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/participant-list-page.ts]: nit: IMO, `assert*` func for visibility are not useful.

File: playwright-e2e/dsm/pages/participant-list-page.ts [hunk 5]
```diff
@@ -335,6 +352,52 @@ export default class ParticipantListPage extends DsmPageBase {
...
+    //Search for a participant that only has an 'OS PE-CGS' cohort tag (shows they registered + enrolled in OS2 only)
+    const participantListTable = this.participantListTable;
+    let amountOfParticipantsDisplayed = await participantListTable.rowsCount;
+
+    let shortId = '';
+    const endTime = Date.now() + 50 * 1000;
+
+    while (shortId === '') {
+      for (let index = 0; index < amountOfParticipantsDisplayed; index++) {
+        const info = await participantListTable.getParticipantDataAt(index, Label.COHORT_TAG_NAME);
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/participant-list-page.ts:407]: Does `getTextAt()` working? If so, no need to do split text.

File: playwright-e2e/dsm/component/tables/participant-list-table.ts [hunk 6]
```diff
@@ -104,4 +136,38 @@ export class ParticipantListTable extends Table {
   private get rowsXPath(): string {
     return '//table/tbody/tr';
   }
+
+  private getDefaultFilterOfStudy(studyName: StudyName): Label[] {
+    const defaultFilterColumnHeaders = [];
+
+    switch (studyName) {
+      case StudyName.RGP:
```
Comment (aweng98) [inline playwright-e2e/dsm/component/tables/participant-list-table.ts]: `tests/dsm/participant-list/rgp-default-filter.spec.ts` covers default filters testing for RGP.

**Comment by Rinchan20** [comment]
others are skipped so I can test out the tests with used saved filters, will turn them back on later

**Comment by Rinchan20** [comment]
test takes ~ 3.3 minutes now (with all tests in test file turned on)

File: playwright-e2e/dsm/component/cohort-tag.ts [hunk 7]
```diff
@@ -44,11 +46,22 @@ export default class CohortTag {
     return `//mat-chip-list//mat-chip[normalize-space(text())="${tagName}"]`;
   }
 
+  public getParticipantPageCohortTagXPathTagFor(tagName: string): string {
+    return `//app-cohort-tag//mat-chip-list//mat-chip[normalize-space(text())='${tagName}']`;
```
Comment (aweng98) [inline playwright-e2e/dsm/component/cohort-tag.ts]: Use double quotes inside the xpath expression.
Also, shorten func name. ParticipantPage string seems to be redundant.

File: playwright-e2e/dsm/component/filters/sections/customize-view.ts [hunk 8]
```diff
@@ -88,6 +99,77 @@ export class CustomizeView {
...
   }
 
+  private useExactName(columnGroupName: ColumnGroup): boolean {
+    if (this.columnGroupsUnableToUseExactName.includes(columnGroupName)) {
+      return false;
+    }
+    return true;
+  }
+
+  public getColumnOption(opts: { columnGroupName: ColumnGroup, groupId: CustomizeViewID, columnOption: Label, instance?: number}): Locator {
```
Comment (aweng98) [inline playwright-e2e/dsm/component/filters/sections/customize-view.ts:109]: I think a new class for Column (a component) would be useful for in including all column* functions. Page object class code would be a lot cleaner.

File: playwright-e2e/dsm/pages/participant-list-page.ts [hunk 9]
```diff
@@ -348,6 +382,52 @@ export default class ParticipantListPage extends DsmPageBase {
...
+      }
+
+      if (shortId === '') {
+        // Get new /filterList response by go to the next page
+        const hasNext = await this._table.paginator.hasNext();
+        if (!hasNext || Date.now() > endTime) {
+          throw new Error(`No participants with only the cohort tag ${tagName} were found`);
+        }
+        await participantListTable.nextPage();
+        amountOfParticipantsDisplayed = await participantListTable.rowsCount;
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/participant-list-page.ts:423]: While loop fix is needed. amountOfParticipantsDisplayed === 0 ==> infinite loop

**Review by aweng98 — APPROVED** [review]

File: playwright-e2e/dsm/component/filters/sections/customize-view.ts (hunk 8 above)
Comment (Rinchan20) [reply playwright-e2e/dsm/component/filters/sections/customize-view.ts:109]: sounds good - will implement in a different PR

**Pull request merged**